moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame.cpp

```diff
@@ -426,131 +437,132 @@ void MotionPlanningFrame::sceneUpdate(planning_scene_monitor::PlanningSceneMonit
+      break;
+    case shapes::SPHERE:
+      shape = std::make_shared<shapes::Sphere>(0.5 * x_length);
+      break;
+    case shapes::CONE:
+      shape = std::make_shared<shapes::Cone>(0.5 * x_length, z_length);
+      break;
+    case shapes::CYLINDER:
+      shape = std::make_shared<shapes::Cylinder>(0.5 * x_length, z_length);
+      break;
+    case shapes::MESH:
+    {
+      QUrl url;
+      if (ui_->shapes_combo_box->currentText().contains("file"))
+        url = QFileDialog::getOpenFileUrl(this, tr("Import Object Mesh"), QString(),
+                                          "CAD files (*.stl *.obj *.dae);;All files (*.*)");
+      else
+        url = QInputDialog::getText(this, tr("Import Object Mesh"), tr("URL for file to import:"), QLineEdit::Normal,
+                                    QString("http://"));
+      if (!url.isEmpty())
+        shape = loadMeshResource(url.toString().toStdString());
+      if (!shape)
         return;
-      }
+      // name mesh objects by their file name
+      selected_shape = url.fileName().toStdString();
+      break;
+    }
+    default:
+      QMessageBox::warning(this, QString("Unsupported shape"),
+                           QString("The '%1' is not supported.").arg(ui_->shapes_combo_box->currentText()));
+  }
 
-      // If the object is very large, ask the user if the scale should be reduced.
-      bool object_is_very_large = false;
-      for (unsigned int i = 0; i < mesh->vertex_count; i++)
-      {
-        if ((abs(mesh->vertices[i * 3 + 0]) > LARGE_MESH_THRESHOLD) ||
-            (abs(mesh->vertices[i * 3 + 1]) > LARGE_MESH_THRESHOLD) ||
-            (abs(mesh->vertices[i * 3 + 2]) > LARGE_MESH_THRESHOLD))
-        {
-          object_is_very_large = true;
-          break;
-        }
-      }
-      if (object_is_very_large)
-      {
-        QMessageBox msg_box;
-        msg_box.setText(
-            "The object is very large (greater than 10 m). The file may be in millimeters instead of meters.");
-        msg_box.setInformativeText("Attempt to fix the size by shrinking the object?");
-        msg_box.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
-        msg_box.setDefaultButton(QMessageBox::Yes);
-        if (msg_box.exec() == QMessageBox::Yes)
-        {
-          for (unsigned int i = 0; i < mesh->vertex_count; ++i)
-          {
-            unsigned int i3 = i * 3;
-            mesh->vertices[i3] *= 0.001;
-            mesh->vertices[i3 + 1] *= 0.001;
-            mesh->vertices[i3 + 2] *= 0.001;
-          }
-        }
-      }
+  // find available (initial) name of object
+  int idx = 0;
+  std::string shape_name = selected_shape + "_" + std::to_string(idx);
+  while (planning_display_->getPlanningSceneRO()->getWorld()->hasObject(shape_name))
+  {
+    idx++;
+    shape_name = selected_shape + "_" + std::to_string(idx);
+  }
 
-      shapes::ShapeConstPtr shape(mesh);
-      Eigen::Isometry3d pose;
-      pose.setIdentity();
+  // Actually add object
```

Comment:
This does not publish the change but only adds the object to the plugin's planning scene copy, right? I would clarify this for first-time readers.
```suggestion
  // Add object to plugin's PlanningScene, but do not publish the change yet (only note that an edit was made)
```